### PR TITLE
Enforce single org for series key reads

### PR DESCRIPTION
**Pull request by stuartcarnie**
_Briefly describe your proposed changes:_

These changes modify the parameters to a series cursor request and render the intent explicitly; series keys returned by the cursor must never cross organization boundaries.

* `SeriesCursorRequest` takes a single `Name` parameter, specifying the 16-byte tsdb encoded org and bucket ID
* `seriesCursor` ensures the first 8 bytes of the measurement name returned from the series index matches the current organization ID

**Blocked by #12010**

### CODE REVIEW

File: storage/series_cursor.go [hunk 1]
```diff
@@ -98,6 +103,9 @@ func (cur *seriesCursor) Next() (*SeriesCursorRow, error) {
 		}
 
 		cur.row.Name, cur.row.Tags = tsdb.ParseSeriesKey(cur.keys[cur.ofs])
+		if !bytes.HasPrefix(cur.row.Name, cur.orgID) {
```
Comment (e-dard) [inline storage/series_cursor.go]: I think we should either change the `orgID` type to something stronger, or if we leave `orgID` as a `[]byte` then maybe `len(cur.row.Name) < 8 || !bytes.Equal(cur.row.Name[:8], cur.orgID)`.

File: storage/series_cursor_test.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+		return tsdb.AppendSeriesKey(nil, name[:], nil)
+	}
+
+	nm := tsdb.EncodeName(0x0f0f, 0xb0b0)
+	cur := &seriesCursor{
+		keys: [][]byte{
+			makeKey(0x0f0f, 0xb0b0),
+			makeKey(0xffff, 0xb0b0),
+		},
+		orgID: nm[:8],
```
Comment (e-dard) [inline storage/series_cursor_test.go]: Can we change `[:8]` to `[:influxdb.OrgIDLength]` or something? I just want to make it simpler to change the id scheme if we need to.

File: storage/series_cursor.go [hunk 3]
```diff
@@ -11,20 +12,28 @@ import (
...
+	Name [16]byte
 }
 
 // seriesCursor is an implementation of SeriesCursor over an tsi1.Index.
 type seriesCursor struct {
 	once  sync.Once
 	index *tsi1.Index
+	sfile *tsdb.SeriesFile
 	mitr  tsdb.MeasurementIterator
+	orgID []byte
```
[the remark this answers is not in the document]
Comment (stuartcarnie) [reply storage/series_cursor.go]: We could assert `len(orgID) == influxdb.IDBytesLen` or similar in `newSeriesCursor` to strengthen this precondition

File: storage/series_cursor.go (hunk 1 above)
Comment (stuartcarnie) [reply storage/series_cursor.go]: `bytes.HasPrefix` is implemented as 

```go
len(name) >= len(orgID) && Equal(name[0:len(prefix)], orgID)
```

So by adding an assertion in `newSeriesCursor`, such as `len(orgID) == influxdb.IDBytesLen`, we could ensure this invariant.

File: storage/series_cursor_test.go (hunk 2 above)
Comment (stuartcarnie) [reply storage/series_cursor_test.go]: Yeah, sounds good; I'll use that constant in the `newSeriesCursor` to validate too

**Comment by stuartcarnie** [comment]
@e-dard, cleaned up the code a bit more:

* replaced `orgID` with `name [influxdb.IDLength]byte` and use that to test the prefix with `name[:influxdb.OrgIDLength]`
* removed all the measurement iterator stuff, as it was redundant

That end-to-end test you shared does exercise this new code and passes, however, there is no easy way to test the negative case, besides explicitly adding ways to break the invariants in the series and index logic.

**Pull request merged**